onnx/ir.h

```diff
@@ -0,0 +1,1048 @@
+  } \
+  const Kind##Attr::ValueType& method(Symbol name) const { \
+    return get<Kind##Attr>(name); \
+  }
+  CREATE_ACCESSOR(Float,f)
+  CREATE_ACCESSOR(Floats,fs)
+  CREATE_ACCESSOR(String,s)
+  CREATE_ACCESSOR(Strings,ss)
+  CREATE_ACCESSOR(Int,i)
+  CREATE_ACCESSOR(Ints,is)
+  CREATE_ACCESSOR(Tensor,t)
+  CREATE_ACCESSOR(Tensors,ts)
+  CREATE_ACCESSOR(Graph,g)
+  CREATE_ACCESSOR(Graphs,gs)
+
+  #undef CREATE_ACCESSOR
+
+private:
+  Derived* This() {
+    return static_cast<Derived*>(this);
+  }
+  template<typename T>
+  Derived* set(Symbol name, typename T::ConstructorType v) {
+    auto it = find(name, false);
+    auto nv = AVPtr(new T(name, std::forward<typename T::ConstructorType>(v)));
+    if(it == values_.end()) {
+      values_.push_back(std::move(nv));
+    } else {
+      *it = std::move(nv);
+    }
+    return This();
+  }
+  template<typename T>
+  typename T::ValueType & get(Symbol name) const {
+    auto it = find(name, true);
+    T* child = dynamic_cast<T*>(it->get());
+    ONNX_ASSERT(child != nullptr);
+    return child->value();
+  }
+  using AVPtr = AttributeValue::Ptr;
+  // NB: For determinism, we use a vector rather than a hash map.  This does
+  // mean that lookups are O(n), so you shouldn't use Attributes to store
+  // a big pile of messages.
+  std::vector<AVPtr> values_;
+  using iterator = std::vector<AVPtr>::iterator;
+  iterator find(Symbol name, bool required) {
+    auto it = std::find_if(values_.begin(), values_.end(),[&](const AVPtr & v) {
+      return v->name == name;
+    });
+    ONNX_ASSERT(!required || it != values_.end());
+    return it;
+  }
+  using const_iterator = std::vector<AVPtr>::const_iterator;
+  const_iterator find(Symbol name, bool required) const {
+    auto it = std::find_if(values_.begin(), values_.end(),[&](const AVPtr & v) {
+      return v->name == name;
+    });
+    ONNX_ASSERTM(!required || it != values_.end(),
+        "%s:%u: %s: required undefined attribute '%s'", __FILE__, __LINE__, __func__, name.toString());
+    return it;
+  }
+};
+
+
+
+// Each use is represented by this type, see Node::uses()
+// 'user' is the consumer of the value, offset is the index into
+// 'user's input this where the produces will be found.
+struct Use {
+  Use(Node * user, size_t offset)
+  : user(user), offset(offset) {}
+  Node * user;
+  size_t offset;
+};
+
+static inline bool operator==(const Use & a, const Use & b) {
+  return a.user == b.user && a.offset == b.offset;
+}
+
+
+// the list types are intentionally simple, but we type-def
+// them here so if we need to change them, refactoring will be easier
+using node_list = std::vector<Node*>;
+using value_list = std::vector<Value*>;
+using use_list = std::vector<Use>;
+using NodeKind = Symbol;
+
+
+struct Value {
+  ONNX_DISALLOW_COPY_AND_ASSIGN(Value);
+  Value(Node * node_, size_t offset_);
+
+private:
+  friend struct Node;
+  friend struct Graph;
+  Node * node_;
+  size_t offset_;
+  size_t unique_ = 0;          // unique id
```

Comment:
The optimizer might need to allocate new names in the process of optimization, and it is a bit more convenient to allocate new numbers than it is to allocate new strings. So that justifies unique id.

In the old design, `uniqueName` was simply the unique id plus an optional debug name (the unique id always uniquely identified.) The ability to set the unique name directly was added in https://github.com/pytorch/pytorch/commit/b97dfc8a92109fc4c311e0767a95ae75ee8ee4a5 to let you say, "please please please make the following node output as some name".

Reply:
Another reason is we need to keep track the input and output names of GraphProto (ValueInfoProto has names). Otherwise after proto->ir->proto converting, we will lose the name information of ValueInfoProto.

Reply:
the API wise is confusing callers, I think. Let's think more on this later.

Reply:
Yeah, I think at least more docs should be helpful.